zkvm/src/verifier.rs

```diff
@@ -55,14 +55,13 @@ impl<'a, 'b> Delegate<r1cs::Verifier<'a, 'b>> for Verifier<'a, 'b> {
         &mut self,
         run: &mut Self::RunType,
     ) -> Result<Option<Instruction>, VMError> {
-        let mut program = Subslice::new_with_range(&run.program, run.offset..run.program.len())?;
-
-        // Reached the end of the program - no more instructions to execute.
-        if program.len() == 0 {
+        let (instr, remainder) = SliceReader::parse(&run.program[run.offset..], |r| {
+            Ok((Instruction::parse(r)?, r.skip_trailing_bytes()))
+        })?;
+        if remainder == 0 {
```

Comment:
Sorry, my mistake. Previous code was checking for length _before_ parsing instruction. The correct way is to: 
1. check that program.offset is not at the end
2. then parse the instruction and update the offset

```rust
// Reached the end of the program - no more instructions to execute.
if run.offset == run.program.len() {
    return Ok(None);
}
let (instr, remainder) = SliceReader::parse(&run.program[run.offset..], |r| {
    Ok((
        Instruction::parse(r)?,
        r.skip_trailing_bytes()
    ))
})?;
run.offset = run.program.len() - remainder;
Ok(Some(instr))
```